src/boot/efi/devicetree.c

```diff
@@ -106,6 +106,118 @@ EFI_STATUS devicetree_install(struct devicetree_state *state, EFI_FILE *root_dir
+                return NULL;
+        const char *strings_block = (const char *) ((const uint8_t *) dt_header + strings_off);
+
+        if (struct_size % sizeof(uint32_t) != 0 ||
+            !ADD_SAFE(&end, struct_off, struct_size) ||
+            end > strings_off)
+                return NULL;
+        size_t size_words = struct_size / sizeof(uint32_t);
+        if (!SUB_SAFE(&end, size_words, 3))
+                return NULL;
+
+        size_t len, name_off, len_words, s;
+
+        for (size_t i = 0; i < end; i++) {
+                switch (be32toh(cursor[i])) {
+                case FDT_BEGIN_NODE:
+                        if (cursor[++i] != 0)
+                                return NULL;
+                        break;
+                case FDT_NOP:
+                        break;
+                case FDT_PROP:
+                        len = be32toh(cursor[++i]);
+                        name_off = be32toh(cursor[++i]);
+                        len_words = DIV_ROUND_UP(len, sizeof(uint32_t));
+
+                        if (ADD_SAFE(&s, name_off, strlen8("compatible")) &&
+                            s < strings_size && streq8(strings_block + name_off, "compatible")) {
+                                const char *c = (const char *) &cursor[++i];
+                                if (len == 0 || i + len_words > size_words || c[len - 1] != '\0')
+                                        c = NULL;
+
+                                return c;
+                        }
+                        i += len_words;
+                        break;
+                default:
+                        return NULL;
+
+                }
+        }
+
+        return NULL;
+}
+
+/* This function checks if the firmware provided DeviceTree
+ * and a UKI provided DeviceTree contain the same first entry
+ * on their respective "compatible" fields. More specifically,
+ * given the FW/UKI "compatible" property pair:
+ *
+ *      compatible = "string1", "string2";
+ *      compatible = "string1", "string3";
+ *
+ * the function reports a match, while for
+ *
+ *      compatible = "string1", "string3";
+ *      compatible = "string2", "string1";
+ *
+ * it reports a mismatch.
+ */
+EFI_STATUS devicetree_match(const void *dtb_buffer, size_t dtb_length) {
+        const void *fw_dtb = find_configuration_table(MAKE_GUID_PTR(EFI_DTB_TABLE));
+        if (!fw_dtb)
+                return EFI_UNSUPPORTED;
+
+        const char *fw_compat = devicetree_get_compatible(fw_dtb);
+
+        return devicetree_match_by_compatible(dtb_buffer, dtb_length, fw_compat);
+}
+
+EFI_STATUS devicetree_match_by_compatible(const void *dtb_buffer, size_t dtb_length, const char *fw_compat) {
+        const FdtHeader *dt_header = ASSERT_PTR(dtb_buffer);
+
+        if (dtb_length < sizeof(FdtHeader) ||
+            dtb_length < be32toh(dt_header->total_size))
+                return EFI_INVALID_PARAMETER;
+
+        if (!fw_compat)
+                return EFI_INVALID_PARAMETER;
```

Comment:
shouldn't we return soe other errror here? EFI_INVALID_PARAMETER sounds a bit EFI_UNSUPPORTEDas if we made a mistake here calling it like this, but that's not really the case, it's a very common case that there's not going to be a dtb table in the firmware (for example on all x86-64)...

maybe return EFI_UNSUPPORTED here.

(or alternatively, leave this func as is, but in devicetree_match() already return EFI_UNSUPPORTED if we can't find the compatible string in the firmware